codegen-server/typescript/src/main/kotlin/software/amazon/smithy/rust/codegen/server/typescript/smithy/generators/TsApplicationGenerator.kt

```diff
@@ -0,0 +1,318 @@
+                    """,
+                    *codegenScope,
+                )
+            }
+            writer.rustBlockTemplate(
+                """pub fn new_socket(address: String, port: i32, backlog: Option<i32>) -> Result<#{socket2}::Socket, Box<dyn std::error::Error>> """.trimIndent(),
+                *codegenScope,
+            ) {
+                writer.rustTemplate(
+                    """
+                    let address: std::net::SocketAddr = format!("{}:{}", address, port).parse()?;
+                    let domain = if address.is_ipv6() {
+                        #{socket2}::Domain::IPV6
+                    } else {
+                        #{socket2}::Domain::IPV4
+                    };
+                    let socket = #{socket2}::Socket::new(domain, #{socket2}::Type::STREAM, Some(#{socket2}::Protocol::TCP))?;
+                    // Set value for the `SO_REUSEPORT` and `SO_REUSEADDR` options on this socket.
+                    // This indicates that further calls to `bind` may allow reuse of local
+                    // addresses. For IPv4 sockets this means that a socket may bind even when
+                    // there's a socket already listening on this port.
+                    socket.set_reuse_port(true)?;
+                    socket.set_reuse_address(true)?;
+                    socket.bind(&address.into())?;
+                    socket.listen(backlog.unwrap_or(1024))?;
+                    Ok(socket)
+                    """.trimIndent(),
+                    *codegenScope,
+                )
+            }
+
+            writer.rust(
+                """
+                /// Clone the inner socket allowing it to be shared between multiple
+                /// Nodejs processes.
+                """.trimIndent(),
+            )
+            Attribute("napi").render(writer)
+            writer.rustBlockTemplate(
+                """pub fn try_clone(&self) -> #{napi}::Result<Self>""".trimIndent(),
+                *codegenScope,
+            ) {
+                writer.rustTemplate(
+                    """
+                    Ok(TsSocket(
+                        self.0
+                            .try_clone()
+                            .map_err(|e| #{napi}::Error::from_reason(e.to_string()))?,
+                    ))
+                    """.trimIndent(),
+                    *codegenScope,
+                )
+            }
+        }
+
+        writer.rustBlock("impl TsSocket") {
+            writer.rustBlockTemplate(
+                """pub fn to_raw_socket(&self) -> #{napi}::Result<#{socket2}::Socket>""".trimIndent(),
+                *codegenScope,
+            ) {
+                writer.rustTemplate(
+                    """
+                    self.0
+                     .try_clone()
+                     .map_err(|e| #{napi}::Error::from_reason(e.to_string()))
+                    """.trimIndent(),
+                    *codegenScope,
+                )
+            }
```

Comment:
Prefer single `rustTemplate` over splitting them like this.

Reply:
I think I'd rather leave it this way for now.